## Coursera Capstone: Segmenting and Clustering Neighborhoods in Toronto
This project is to explore, segment, and cluster neighborhoods in the city of Toronto.

## Introduction

We will start by scraping the Wikipedia page <a href="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M">List of postal codes of Canada: M</a> to get a list of neighborhoods in Toronto. Then we will find the latitude and longitude of each neighborhood. Next, we will use the Foursquare API to explore the neighborhoods. We will get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. We will use the *k*-means clustering algorithm to complete this task. Finally, we will use the Folium library to visualize the neighborhoods in Toronto and their emerging clusters.

## Table of Contents

1. <a href="#toc1">Acquire Data</a>

2. <a href="#toc2">Explore Neighborhoods in Toronto</a>

3. <a href="#toc3">Analyze Each Neighborhood</a>

4. <a href="#toc4">Cluster Neighborhoods</a>

5. <a href="#toc5">Examine Clusters</a>    


First download all the dependencies that we will need for this analysis.

In [18]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analysis

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id="toc1"></a>
## 1. Acquire Data

We will scrape the list of Toronto neighborhoods from the Wikipedia page <a href="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M">List of postal codes of Canada: M</a>. Postal codes beginning with M are located within the city of Toronto. Then we will find the geographical coordinates of each post code.

#### Read data from web page

We will read the page and convert the table into a dataframe using the *pandas* read_html method.

In [2]:
#get Toronto neighborhoods from Wikipedia page
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#read the Wikipedia page - returns list of dataframes
dfs = pd.read_html(url, header=0)
#take the first dataframe from the returned list (it should be the only dataframe in the list)
df = dfs[0]
df.head(10)

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront
5      M5A  Downtown Toronto       Regent Park
6      M6A        North York  Lawrence Heights
7      M6A        North York    Lawrence Manor
8      M7A      Queen's Park      Not assigned
9      M8A      Not assigned      Not assigned

We will ignore any records where *Borough* is "Not assigned".

In [3]:
#create new dataframe with records where Borough is not 'Not assigned'
df_assigned = df[df['Borough'] != 'Not assigned']
df_assigned.head(10)

Postcode           Borough     Neighbourhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Not assigned
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge
12      M1B       Scarborough           Malvern

In rows where the value of *Neighbourhood* is 'Not assigned' we will replace it with the value of the *Borough*.

In [4]:
#create a list of neighborhoods, replacing the borough where neighborhood is 'Not assigned'
new_neigh = df_assigned['Neighbourhood'].where(df_assigned['Neighbourhood'] != 'Not assigned', other = df_assigned['Borough'], axis = 0)
#construct new dataframe using postcode and borough from the previous dataframe and neighborhood from the above list
df_replaced = pd.concat([df_assigned['Postcode'], df_assigned['Borough'], new_neigh], axis = 1)
df_replaced.head(10)

Postcode           Borough     Neighbourhood
2       M3A        North York         Parkwoods
3       M4A        North York  Victoria Village
4       M5A  Downtown Toronto      Harbourfront
5       M5A  Downtown Toronto       Regent Park
6       M6A        North York  Lawrence Heights
7       M6A        North York    Lawrence Manor
8       M7A      Queen's Park      Queen's Park
10      M9A         Etobicoke  Islington Avenue
11      M1B       Scarborough             Rouge
12      M1B       Scarborough           Malvern

Whenever we have more than one row per postcode, we will concatenate all neighborhoods into a comma separated list

In [5]:
#group the dataframe by Postcode and Borough and concatenate all neighborhoods into comma separated list
toronto_neighborhoods = df_replaced.groupby(['Postcode', 'Borough'])['Neighbourhood'].apply(list).apply(lambda x: ', '.join(x)).to_frame()
toronto_neighborhoods.reset_index(inplace = True)
toronto_neighborhoods.head(10)

Postcode      Borough                                    Neighbourhood
0      M1B  Scarborough                                   Rouge, Malvern
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough                Guildwood, Morningside, West Hill
3      M1G  Scarborough                                           Woburn
4      M1H  Scarborough                                        Cedarbrae
5      M1J  Scarborough                              Scarborough Village
6      M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park
7      M1L  Scarborough                  Clairlea, Golden Mile, Oakridge
8      M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West
9      M1N  Scarborough                      Birch Cliff, Cliffside West

In [6]:
#display the shape of the resulting dataframe
toronto_neighborhoods.shape

(103, 3)

### Assign geographical coordinates

The next step is to assign the geographical coordinates for each postcode. Instructions on the Coursera submission page suggest to use the Geocoder Python package. Unfortunately I was not able to get usable results from this package. As suggested on the submission page, I decided to use the dataset <a href='http://cocl.us/Geospatial_data'>http://cocl.us/Geospatial_data</a> with predefined coordinates per postal code.  

There is an alternate way by which we could get coordinates and that would be to use the <a href='https://geopy.readthedocs.io/en/stable/'>GeoPy</a> library with the Nominatim geolocator service (instead of Google Maps). However, this library does not return coordinates based on postal codes but rather on neighborhood names. This would mean that we would have to restructure the above dataframe back to what it was before we concatenated neighborhoods into comma separated values. I decided not to go this way, although I suspect it would have been a viable approach to solving this exercise.

In [7]:
#read provided dataset
coords = pd.read_csv('http://cocl.us/Geospatial_data')
coords.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We will merge this dataset with our *toronto_neighborhoods* dataset from above.

In [8]:
toronto_neighborhoods = pd.merge(toronto_neighborhoods, coords, left_on = 'Postcode', right_on = 'Postal Code')[['Postcode', 'Borough', 'Neighbourhood', 'Latitude', 'Longitude']]
toronto_neighborhoods.head(10)

Postcode      Borough                                    Neighbourhood  \
0      M1B  Scarborough                                   Rouge, Malvern   
1      M1C  Scarborough           Highland Creek, Rouge Hill, Port Union   
2      M1E  Scarborough                Guildwood, Morningside, West Hill   
3      M1G  Scarborough                                           Woburn   
4      M1H  Scarborough                                        Cedarbrae   
5      M1J  Scarborough                              Scarborough Village   
6      M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park   
7      M1L  Scarborough                  Clairlea, Golden Mile, Oakridge   
8      M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West   
9      M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

In [9]:
toronto_neighborhoods.shape

(103, 5)

<a id="toc2"></a>
## 2. Explore Neighborhoods in Toronto

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [10]:
# The code was removed by Watson Studio for sharing.

#### Let's explore the first neighborhood in our dataframe

Find the name, latitude and longitude of the first neighborhood in the dataframe.

In [11]:
neighborhood_name = toronto_neighborhoods.loc[0, 'Neighbourhood'] # neighborhood name
neighborhood_latitude = toronto_neighborhoods.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_neighborhoods.loc[0, 'Longitude'] # neighborhood longitude value

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rouge, Malvern are 43.806686299999996, -79.19435340000001.


#### Now, let's get the top 100 venues that are in the above neighborhood within a radius of 500 meters

First we will create the GET request URL.

In [12]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)


Send the GET request and examine the results

In [13]:
results = requests.get(url).json()
#results.head()

We see that all the information that we want is in the *items* key. Before we proceed, we will create the **get_category_type** function which extracts the category name from a JSON object.

In [14]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [20]:
#extract the items key from the results
venues = results['response']['groups'][0]['items']
#flatten JSON into a dataframe
nearby_venues = json_normalize(venues) 
#filter columns that we need for further analysis
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]
#extract the category for each row using the previously defined function
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)
# clean columns ?????????????????????????????????????????????
#nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

venue.name      venue.categories  venue.location.lat  venue.location.lng
0    Wendy's  Fast Food Restaurant           43.807448          -79.199056

#### Create a function to repeat the same process as above to all the neighborhoods in Toronto

In [16]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

The following code executes the above function on each neighborhood and creates a new dataframe called *toronto_venues*.

In [21]:
toronto_venues = getNearbyVenues(names = toronto_neighborhoods['Neighbourhood'],
                                   latitudes = toronto_neighborhoods['Latitude'],
                                   longitudes = toronto_neighborhoods['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West, Steeles West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The D

Check the size and first few rows of the resulting dataframe

In [22]:
print(toronto_venues.shape)
toronto_venues.head()

(2228, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2       Guildwood, Morningside, West Hill              43.763573   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3              -79.188711                G & G Electronics       43.765309   
4              -79.188711                 Big Bite Burrito       43.766299   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.190720    Mexican Restaurant

Count number of venue categories by neighborhood

In [23]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                      9   
Alderwood, Long Branch                                                  8   
Bathurst Manor, Downsview North, Wilson Heights                        17   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      25   
Berczy Park                                                            56   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      6   
Brockton, Exhibition Place, Parkdale Village                           18   
Business Reply Mail Processing Centre 969 Eastern                      18   
CFB Toronto, Downsview East                                             3   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            46   
Caledonia-Fairbanks                                                     6   
Canada Post Gateway Processing Centre                                  10   
Cedarbrae                                                               7   
Central Bay Street                                                     82   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   86   
Clairlea, Golden Mile, Oakridge                                        10   
Clarks Corners, Sullivan, Tam O'Shanter                                10   
Cliffcrest, Cliffside, Scarborough Village West                         2   
Cloverdale, Islington, Martin Grove, Princess G...                      1   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             39   
Davisville North                                                        8   
...                                                                   ...   
Northwest                                                               3   
Northwood Park, York University                                         4   
Parkdale, Roncesvalles                                                 16   
Parkwoods                                                               3   
Queen's Park                                                           45   
Rosedale                                                                4   
Roselawn                                                                1   
Rouge, Malvern                                                          1   
Runnymede, Swansea                                                     37   
Ryerson, Garden District                                              100   
Scarborough Village                                                     2   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        93   
Studio District                                                        39   
The Annex, North Midtown, Yorkville                                    24   
The Beaches                                                             4   
The Beaches West, India Bazaar                                         19   
The Danforth West, Riverdale                                           42   
The Junction North, Runnymede                                           4   
The Kingsway, Montgomery Road, Old Mill North   

In [24]:
print('There are {} unique categories.'.format(len(toronto_venues['Venue Category'].unique())))


There are 275 unique categories.


We see that many neighborhoods have fewer than 10 different venue categories. It doesn't make sense to include these in segmentation because there is not enough data to characterize them. Therefore we will exclude these from the dataset.  

But first, we have to check how many neighborhoods have fewer than 10 venue categories to ensure that we still have enough neighborhoods left for segmentation.

In [25]:
#store the results of the above counts into a dataframe
toronto_venues_count = toronto_venues.groupby('Neighborhood').count()
print('There are {} neighborhoods with fewer than 10 venue categories.'.format(toronto_venues_count[toronto_venues_count['Venue Category'] < 10].shape[0]))

There are 51 neighborhoods with fewer than 10 venue categories.


We know that we have a total of 103 neighborhoods which means that even after we remove 51 of them we should still have sufficient data for segmentation.  

Therefore we will exclude the neighborhoods with fewer than 10 venue categories.

In [26]:
#create list with neighborhoods to exclude
neigh_to_exclude = toronto_venues_count[toronto_venues_count['Venue Category'] < 10].index.tolist()
#create filtered dataframe by excluding neighborhoods in above list
toronto_venues_filt = toronto_venues[~toronto_venues['Neighborhood'].isin(neigh_to_exclude)]
#check size of resulting dataframe
toronto_venues_filt.groupby('Neighborhood').count().shape

(48, 6)

The number of neighborhoods is sufficient for further analysis. We will rename the filtered dataset *toronto_venues_filt* back to the original dataset *toronto_venues*.

In [27]:
toronto_venues = toronto_venues_filt

<a id="toc3"></a>
## 3. Analyze Each Neighborhood

We will do one hot encoding to pivot category values into columns of the dataframe.  

There is one observation that we have to be careful about: one of the category values is *Neighborhood*. After one hot encoding, this value will become a column name. We are already using the column *Neighborhood* to represent the neighborhood name. To avoid confusing these columns, we will rename the column that comes from one hot encoding as *Neighborhood Category*.

In [28]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

#rename the column 'Neighborhood' which represents a category name to 'Neighborhood Category' 
#this is to distinguish this column from the 'Neighborhood' column which we want to continue to use as the neighborhood name
toronto_onehot.rename(columns={'Neighborhood':'Neighborhood Category'}, inplace=True)

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

#toronto_onehot.head()


Check the new dataframe size:

In [29]:
toronto_onehot.shape

(2024, 258)

#### We will group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [30]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head(10)

Neighborhood  Accessories Store  \
0                           Adelaide, King, Richmond                0.0   
1    Bathurst Manor, Downsview North, Wilson Heights                0.0   
2                  Bedford Park, Lawrence Manor East                0.0   
3                                        Berczy Park                0.0   
4       Brockton, Exhibition Place, Parkdale Village                0.0   
5  Business Reply Mail Processing Centre 969 Eastern                0.0   
6  CN Tower, Bathurst Quay, Island airport, Harbo...                0.0   
7                        Cabbagetown, St. James Town                0.0   
8              Canada Post Gateway Processing Centre                0.0   
9                                 Central Bay Street                0.0   

   Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0             0.0                0.0  0.000000            0.000000   
1             0.0                0.0  0.000000            0.000000   
2             0.0                0.0  0.000000            0.000000   
3             0.0                0.0  0.000000            0.000000   
4             0.0                0.0  0.000000            0.000000   
5             0.0                0.0  0.000000            0.000000   
6             0.0                0.0  0.071429            0.071429   
7             0.0                0.0  0.000000            0.000000   
8             0.0                0.0  0.000000            0.000000   
9             0.0                0.0  0.000000            0.000000   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0      0.000000        0.000000         0.000000          0.000000   
1      0.000000        0.000000         0.000000          0.000000   
2      0.000000        0.000000         0.000000          0.000000   
3      0.000000        0.000000         0.000000          0.000000   
4      0.000000        0.000000         0.000000          0.000000   
5      0.000000        0.000000         0.000000          0.000000   
6      0.071429        0.142857         0.142857          0.142857   
7      0.000000        0.000000         0.000000          0.000000   
8      0.000000        0.000000         0.000000          0.000000   
9      0.000000        0.000000         0.000000          0.000000   

      ...       Train Station  Vegetarian / Vegan Restaurant  \
0     ...                 0.0                       0.010000   
1     ...                 0.0                       0.000000   
2     ...                 0.0                       0.000000   
3     ...                 0.0                       0.000000   
4     ...                 0.0                       0.000000   
5     ...                 0.0                       0.000000   
6     ...                 0.0                       0.000000   
7     ...                 0.0                       0.000000   
8     ...                 0.0                       0.000000   
9     ...                 0.0                       0.012195   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0               0.0     0.000000                    0.0              0.0   
1               0.0     0.058824                    0.0              0.0   
2               0.0     0.000000                    0.0              0.0   
3               0.0     0.000000                    0.0              0.0   
4               0.0     0.000000                    0.0              0.0   
5               0.0     0.000000                    0.0              0.0   
6               0.0     0.000000                    0.0              0.0   
7               0.0     0.000000                    0.0              0.0   
8               0.0     0.000000                    0.0              0.0   
9               0.0     0.000000                    0.0              0.0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0  0.010000          0.0           0.01     0.000000  
1  0.000000         

Check the new dataframe size:

In [31]:
toronto_grouped.shape

(48, 258)

#### Store the above into a *pandas* dataframe

Write a function to sort the venues in descending order

In [32]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Create the new dataframe and display the top 10 venues for each neighborhood

In [33]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns = columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()


Neighborhood 1st Most Common Venue  \
0                         Adelaide, King, Richmond           Coffee Shop   
1  Bathurst Manor, Downsview North, Wilson Heights           Coffee Shop   
2                Bedford Park, Lawrence Manor East           Coffee Shop   
3                                      Berczy Park           Coffee Shop   
4     Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant   American Restaurant   
1    Frozen Yogurt Shop           Bridal Shop      Sushi Restaurant   
2    Italian Restaurant  Fast Food Restaurant           Pizza Place   
3          Cocktail Bar            Restaurant    Seafood Restaurant   
4                  Café           Coffee Shop             Pet Store   

  5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
0            Steakhouse                    Gym        Clothing Store   
1   Fried Chicken Joint             Restaurant                  Bank   
2              Pharmacy             Restaurant               Butcher   
3                Bakery             Steakhouse                   Pub   
4          Climbing Gym  Performing Arts Venue         Burrito Place   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Asian Restaurant                Bakery                    Bar  
1                 Diner        Sandwich Place          Deli / Bodega  
2                  Café                   Pub          Grocery Store  
3                  Café           Cheese Shop               Beer Bar  
4               Stadium                   Bar     Italian Restaurant

In [34]:
neighborhoods_venues_sorted.groupby(['1st Most Common Venue']).size()

1st Most Common Venue
Accessories Store       1
Airport Terminal        1
Bar                     1
Breakfast Spot          2
Bus Line                1
Café                    6
Clothing Store          1
Coffee Shop            21
Greek Restaurant        1
Gym                     2
Indian Restaurant       1
Japanese Restaurant     1
Light Rail Station      1
Pharmacy                1
Pizza Place             2
Restaurant              1
Sandwich Place          2
Skating Rink            1
Social Club             1
dtype: int64

<a id="toc4"></a>
## 4. Cluster Neighborhoods


Run *k*-means to cluster the neighborhood into 5 clusters.

In [35]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([2, 3, 3, 2, 0, 3, 4, 3, 2, 2, 0, 0, 2, 1, 3, 2, 3, 2, 2, 0, 2, 2, 2,
       0, 2, 2, 0, 3, 2, 2, 0, 2, 0, 2, 0, 2, 3, 2, 2, 2, 0, 3, 3, 3, 3, 2,
       3, 3], dtype=int32)

Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [36]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = neighborhoods_venues_sorted

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(toronto_neighborhoods.set_index('Neighbourhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

Cluster Labels                                     Neighborhood  \
0               2                         Adelaide, King, Richmond   
1               3  Bathurst Manor, Downsview North, Wilson Heights   
2               3                Bedford Park, Lawrence Manor East   
3               2                                      Berczy Park   
4               0     Brockton, Exhibition Place, Parkdale Village   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                  Café       Thai Restaurant   
1           Coffee Shop    Frozen Yogurt Shop           Bridal Shop   
2           Coffee Shop    Italian Restaurant  Fast Food Restaurant   
3           Coffee Shop          Cocktail Bar            Restaurant   
4        Breakfast Spot                  Café           Coffee Shop   

  4th Most Common Venue 5th Most Common Venue  6th Most Common Venue  \
0   American Restaurant            Steakhouse                    Gym   
1      Sushi Restaurant   Fried Chicken Joint             Restaurant   
2           Pizza Place              Pharmacy             Restaurant   
3    Seafood Restaurant                Bakery             Steakhouse   
4             Pet Store          Climbing Gym  Performing Arts Venue   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0        Clothing Store      Asian Restaurant                Bakery   
1                  Bank                 Diner        Sandwich Place   
2               Butcher                  Café                   Pub   
3                   Pub                  Café           Cheese Shop   
4         Burrito Place               Stadium                   Bar   

  10th Most Common Venue Postcode           Borough   Latitude  Longitude  
0                    Bar      M5H  Downtown Toronto  43.650571 -79.384568  
1          Deli / Bodega      M3H        North York  43.754328 -79.442259  
2          Grocery Store      M5M        North York  43.733283 -79.419750  
3               Beer Bar      M5E  Downtown Toronto  43.644771 -79.373306  
4     Italian Restaurant      M6K      West Toronto  43.636847 -79.428191

Visualize the resulting clusters

In [37]:
address = 'Toronto'
geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude


In [38]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Note: because GitHub doesn't display Folium maps, a print screen of the map is available <a href='img/Toronto.png'>here</a>.

<a id="toc5"></a>
## 5. Examine Clusters


We will examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, we can then assign a name to each cluster.

### Cluster 0: Residential

In [39]:
toronto_cluster0 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]-2))]]
toronto_cluster0

Neighborhood 1st Most Common Venue  \
4   Brockton, Exhibition Place, Parkdale Village        Breakfast Spot   
10     Chinatown, Grange Park, Kensington Market                  Café   
11                                      Christie                  Café   
19                  Dovercourt Village, Dufferin              Pharmacy   
23                Harbord, University of Toronto                  Café   
26                 High Park, The Junction South                  Café   
30              Lawrence Heights, Lawrence Manor     Accessories Store   
32                      Little Portugal, Trinity                   Bar   
34                        Parkdale, Roncesvalles        Breakfast Spot   
40                               Studio District                  Café   

   2nd Most Common Venue          3rd Most Common Venue  \
4                   Café                    Coffee Shop   
10                   Bar  Vegetarian / Vegan Restaurant   
11         Grocery Store                           Park   
19        Discount Store                         Bakery   
23            Restaurant                            Bar   
26    Mexican Restaurant                      Gastropub   
30        Clothing Store                       Boutique   
32           Men's Store                           Café   
34             Gift Shop             Italian Restaurant   
40           Coffee Shop             Italian Restaurant   

        4th Most Common Venue 5th Most Common Venue   6th Most Common Venue  \
4                   Pet Store          Climbing Gym   Performing Arts Venue   
10      Vietnamese Restaurant           Coffee Shop                  Bakery   
11                 Restaurant             Nightclub      Italian Restaurant   
19                Supermarket  Gym / Fitness Center  Furniture / Home Store   
23        Japanese Restaurant           Coffee Shop                  Bakery   
26  Cajun / Creole Restaurant                Bakery      Italian Restaurant   
30                 Shoe Store    Miscellaneous Shop  Furniture / Home Store   
32                Coffee Shop            Restaurant        Asian Restaurant   
34               Burger Joint            Restaurant               Bookstore   
40                     Bakery   American Restaurant                   Diner   

    7th Most Common Venue 8th Most Common Venue      9th Most Common Venue  \
4           Burrito Place               Stadium                        Bar   
10    Dumpling Restaurant    Mexican Restaurant         Chinese Restaurant   
11     Athletics & Sports                 Diner                 Baby Store   
19                   Pool           Music Venue                       Café   
23              Bookstore        Sandwich Place                  Jazz Club   
26    Fried Chicken Joint                 Diner                  Speakeasy   
30              Gift Shop           Event Space                Coffee Shop   
32  Vietnamese Restaurant                Bakery               Cocktail Bar   
34          Movie Theater                  Bank               Dessert Shop   
40            Fish Market             Juice Bar  Latin American Restaurant   

       10th Most Common Venue Postcode           Borough  
4          Italian Restaurant      M6K      West Toronto  
10                Gaming Cafe      M5T  Downtown Toronto  
11                Coffee Shop      M6G  Downtown Toronto  
19  Middle Eastern Restaurant      M6H      West Toronto  
23                   Beer Bar      M5S  Downtown Toronto  
26     Furniture / Home Store      M6P      West Toronto  
30      Vietnamese Restaurant      M6A        North York  
32                Pizza Place      M6J      West Toronto  
34                        Bar      M6R      West Toronto  
40                  Bookstore      M4M      East Toronto

The venue categories in this cluster are predominantly shops with some interspersed coffee shops and restaurants as well as parks and sporting venues. It apears that this cluster represents residential areas.

### Cluster 1: Ground transportation hub

In [40]:
toronto_cluster1 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]-2))]]
toronto_cluster1

Neighborhood 1st Most Common Venue  \
13  Clairlea, Golden Mile, Oakridge              Bus Line   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13                Bakery  Fast Food Restaurant         Metro Station   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13          Intersection           Bus Station          Soccer Field   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  \
13                  Park  Eastern European Restaurant    Dumpling Restaurant   

   Postcode      Borough  
13      M1L  Scarborough

The venues in this cluster suggest that it is a transportation hub as it includes bus and metro stations.

### Cluster 2: Downtown

In [41]:
toronto_cluster2 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]-2))]]
toronto_cluster2

Neighborhood 1st Most Common Venue  \
0                            Adelaide, King, Richmond           Coffee Shop   
3                                         Berczy Park           Coffee Shop   
8               Canada Post Gateway Processing Centre           Coffee Shop   
9                                  Central Bay Street           Coffee Shop   
12                               Church and Wellesley   Japanese Restaurant   
15                     Commerce Court, Victoria Hotel           Coffee Shop   
17  Deer Park, Forest Hill SE, Rathnelly, South Hi...           Coffee Shop   
18           Design Exchange, Toronto Dominion Centre           Coffee Shop   
20                       Fairview, Henry Farm, Oriole        Clothing Store   
21             First Canadian Place, Underground city                  Café   
22                   Flemingdon Park, Don Mills South                   Gym   
24  Harbourfront East, Toronto Islands, Union Station           Coffee Shop   
25                          Harbourfront, Regent Park           Coffee Shop   
28  Kingsway Park South West, Mimico NW, The Queen...           Social Club   
29                      L'Amoreaux West, Steeles West           Coffee Shop   
31                                            Leaside           Coffee Shop   
33                                 North Toronto West           Coffee Shop   
35                                       Queen's Park           Coffee Shop   
37                           Ryerson, Garden District           Coffee Shop   
38                                     St. James Town           Coffee Shop   
39                    Stn A PO Boxes 25 The Esplanade           Coffee Shop   
45                                   Willowdale South            Restaurant   

   2nd Most Common Venue      3rd Most Common Venue  \
0                   Café            Thai Restaurant   
3           Cocktail Bar                 Restaurant   
8                  Hotel  Middle Eastern Restaurant   
9                   Café         Italian Restaurant   
12      Sushi Restaurant                Coffee Shop   
15                  Café                 Restaurant   
17                   Pub         Light Rail Station   
18                 Hotel                       Café   
20  Fast Food Restaurant                Coffee Shop   
21           Coffee Shop                      Hotel   
22      Asian Restaurant                Coffee Shop   
24                 Hotel                   Aquarium   
25                  Park                     Bakery   
28        Sandwich Place                Wings Joint   
29    Chinese Restaurant       Fast Food Restaurant   
31   Sporting Goods Shop              Grocery Store   
33   Sporting Goods Shop              Grocery Store   
35                 Diner        Japanese Restaurant   
37        Clothing Store                       Café   
38            Restaurant                      Hotel   
39            Restaurant                       Café   
45      Ramen Restaurant                       Café   

        4th Most Common Venue 5th Most Common Venue  \
0         American Restaurant            Steakhouse   
3          Seafood Restaurant                Bakery   
8    Mediterranean Restaurant   Fried Chicken Joint   
9                Burger Joint                   Bar   
12                    Gay Bar          Burger Joint   
15                      Hotel   American Restaurant   
17          Convenience Store      Sushi Restaurant   
18                 Restaurant   American Restaurant   
20                 Restaurant         Women's Store   
21                 Restaurant   American Restaurant   
22                 Beer Store    Italian Restaurant   
24                Pizza Place                  Café   
25                       Café                   Pub   
28       Fast Food Restaurant                   Gym   
29                Pizza Place     Indian Restaurant   
31               Burger Joint          Liquor Store   
33       Fast Food Res

Most of the neighborhoods in this cluster appear to be geographically located near downtown. The venue categories are predominantly restaurants, coffee shops and hotels with some shops and yoga studios. This is representative of a *Downtown* city area.

### Cluster 3: Quick eats

In [42]:
toronto_cluster3 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]-2))]]
toronto_cluster3

Neighborhood 1st Most Common Venue  \
1     Bathurst Manor, Downsview North, Wilson Heights           Coffee Shop   
2                   Bedford Park, Lawrence Manor East           Coffee Shop   
5   Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
7                         Cabbagetown, St. James Town           Coffee Shop   
14            Clarks Corners, Sullivan, Tam O'Shanter           Pizza Place   
16                                         Davisville        Sandwich Place   
27       Humber Bay Shores, Mimico South, New Toronto                   Gym   
36                                 Runnymede, Swansea           Coffee Shop   
41                The Annex, North Midtown, Yorkville           Coffee Shop   
42                     The Beaches West, India Bazaar        Sandwich Place   
43                       The Danforth West, Riverdale      Greek Restaurant   
44                                   Thorncliffe Park     Indian Restaurant   
46                    Woodbine Gardens, Parkview Hill           Pizza Place   
47                                   Woodbine Heights          Skating Rink   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1     Frozen Yogurt Shop           Bridal Shop      Sushi Restaurant   
2     Italian Restaurant  Fast Food Restaurant           Pizza Place   
5            Yoga Studio         Garden Center            Skate Park   
7                   Park            Restaurant    Italian Restaurant   
14    Italian Restaurant   Rental Car Location       Thai Restaurant   
16          Dessert Shop           Pizza Place    Seafood Restaurant   
27   American Restaurant          Liquor Store  Fast Food Restaurant   
36                  Café           Pizza Place      Sushi Restaurant   
41                  Café        Sandwich Place           Pizza Place   
42                  Park  Fast Food Restaurant            Steakhouse   
43           Coffee Shop        Ice Cream Shop             Bookstore   
44           Yoga Studio           Supermarket         Grocery Store   
46  Fast Food Restaurant  Gym / Fitness Center             Gastropub   
47              Pharmacy        Cosmetics Shop            Beer Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1    Fried Chicken Joint            Restaurant                  Bank   
2               Pharmacy            Restaurant               Butcher   
5                   Park               Brewery        Farmers Market   
7                 Bakery                  Café                   Pub   
14          Noodle House    Chinese Restaurant  Fast Food Restaurant   
16      Sushi Restaurant    Italian Restaurant            Restaurant   
27    Seafood Restaurant    Mexican Restaurant        Sandwich Place   
36    Italian Restaurant              Pharmacy        Cosmetics Shop   
41              Pharmacy     Indian Restaurant        Cosmetics Shop   
42      Sushi Restaurant     Food & Drink Shop                   Pub   
43    Italian Restaurant           Yoga Studio        Cosmetics Shop   
44          Liquor Store                  Park              Pharmacy   
46                  Café    Rock Climbing Spot             Pet Store   
47              Bus Stop    Athletics & Sports          Dance Studio   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
1                  Diner        Sandwich Place          Deli / Bodega   
2                   Café                   Pub          Grocery Store   
5   Fast Food Restaurant         Burrito Place             Restaurant   
7            Pizza Place              Pharmacy         Breakfast Spot   
14              Pharmacy   Fried Chicken Joint             Donut Shop   
16                  Café           Coffee Shop                   Park   
27                Bakery            Restaurant    Fried Chicken Joint   
36             Bookstore        Sandwich Place          Burrito Place   
41                   Pub          Bu

The venues in this cluster are predominantly restaurants/fast food restaurants and similar and this suggests that it is an area for grabbing something quick to eat.

### Cluster 4: Airport

In [44]:
toronto_cluster4 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(2, toronto_merged.shape[1]-2))]]
toronto_cluster4

Neighborhood 1st Most Common Venue  \
6  CN Tower, Bathurst Quay, Island airport, Harbo...      Airport Terminal   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6       Airport Service        Airport Lounge                 Plane   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6          Airport Gate      Sculpture Garden         Boat or Ferry   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue Postcode  \
6    Airport Food Court               Airport               Boutique      M5V   

            Borough  
6  Downtown Toronto

This cluster is obviously an airport.